File: heagen/lib/xronos/fits/xrftgdky.c
/*
 * xrftgdky.c
 */
#include "fitsio.h"

int xrftgdky(fitsfile *fptr, char *keynami, char *keynamf, char *keynams,
             double *dval, int *status) {
/*
 * XRonos FiTs routine to Get a Double KeYword value.
 * 
 * This general-purpose routine could actually be a fitsio routine.
 * All it does is to allow the possibilty that a double keyword value
 * may be represented in a file as two keywords: one integer and one
 * double precision.  
 * 
 * First, the routine looks for the integer-double pair given by
 * keynami and keynamf, respectively.  If it finds those keywords,
 * it returns their sum in dval.  If the pair are not found, it looks
 * for the single keyword given by keynams and returns its value in dval.
 * 
 *  I  fptr          = Pointer to fitsfile
 *  I  keynami  (c)  = Name of the integer part of the keyword
 *  I  keynamf  (c)  = Name of the fractional part of the keyword
 *  I  keynams  (c)  = Name of the single keyword to use if double is not there
 *  O  dval     (d)  = Returned double precision value [dble(i) + f]
 *  O  status   (i)  = 0 if succesful.
 */
   double intpart, decpart;
/*
 * Search for paired keyword first.
 */
   *status = 0;
   fits_read_key_dbl(fptr, keynami, &intpart, NULL, status);
   if ( *status == 0 ) {
      fits_read_key_dbl(fptr, keynamf, &decpart, NULL, status);
      if ( *status == 0 ) {
         *dval = intpart + decpart;
         return(*status);
      }
   }
/*
 * Search for single keyword.
 */
   *status = 0;
   fits_read_key_dbl(fptr, keynams, dval, NULL, status);
   return(*status);
}

File: ftools/timing/xronos/xronlib/fits/xrftgdky.f
      subroutine xrftgdky(lui,keynami,keynamf,keynams,dval,ftstat)
      implicit none

c XRonos FiTs routine to Get a Double KeYword value.

c This general-purpose routine could actually be a fitsio routine.
c All it does is to allow the possibilty that a double keyword value
c may be represented in a file as two keywords: one integer and one
c double precision.  

c First, the routine looks for the integer-double pair given by
c keynami and keynamf, respectively.  If it finds those keywords,
c it returns their sum in dval.  If the pair are not found, it looks
c for the single keyword given by keynams and returns its value in dval.

c  I  lui      (i)  = lu of input fits file
c  I  keynami  (c)  = Name of the integer part of the keyword
c  I  keynamf  (c)  = Name of the fractional part of the keyword
c  I  keynams  (c)  = Name of the single keyword in case the double is not there
c  O  dval     (d)  = Returned double precision value [dble(i) + f]
c  O  ftstat   (i)  = 0 if succesful.

c Subroutines called: ftgkyj, ftgkyd

c Author: EAL, HEASARC/GSFC  HSTX, April 1994

      character*(*) keynami,keynamf,keynams
      character(16) comm
      integer lui,ival,ftstat
      double precision dval

      ftstat = 0

c Search for paired keyword first.

      CALL    ftgkyj(lui,keynami,ival,comm,ftstat)
      if(ftstat.eq.0) then
         CALL ftgkyd(lui,keynamf,dval,comm,ftstat)
         dval = dble(ival) + dval
      else

c Search for single keyword.

         ftstat = 0
         CALL ftgkyd(lui,keynams,dval,comm,ftstat)
      endif

      return
      end
